### PR TITLE
Add an option to set a destination directory for the simulation output (local)

### PR DESCRIPTION
**Pull request by JoschD**
In this tiny PR:
 - Added `output_destination` option (thank you @anabramo, as in #33 )
 - Fixed doc
 - Restructured and documented code
 - added tests
 - bumped major version, as I am breaking all kinds of things
 
 
 Could this PR have been 4 smaller PRs? Maybe. 
 But where is the fun in that?

### CODE REVIEW

File: pylhc_submitter/job_submitter.py [hunk 1]
```diff
@@ -273,214 +326,18 @@ def main(opt):
...
-        opt.append_jobs,
-        opt.executable,
-        opt.script_arguments,
-        opt.script_extension,
-    )
-    job_df, dropped_jobs = _drop_already_ran_jobs(
-        job_df, opt.resume_jobs or opt.append_jobs, opt.job_output_dir, opt.check_files
-    )
+    save_config(Path(opt.working_directory), opt, "job_submitter")
+    creation_opt, runner_opt = check_opts(opt)
```
Comment (fsoubelet) [inline pylhc_submitter/job_submitter.py:330]: Why are we saving the config before checking options now?

File: pylhc_submitter/job_submitter.py [hunk 2]
```diff
@@ -519,25 +382,24 @@ def _check_opts(opt):
...
-        LOG.info(job_name)
-    LOG.info("--------- JOBS FINISHED: NAMES ------------")
-    for job_name in finished_jobs:
-        LOG.info(job_name)
+    
+    # Create new classes
+    opt.output_dir = opt.job_output_dir  # renaming
+
+    creation = CreationOpts(**{f.name: opt[f.name] for f in fields(CreationOpts)})
+    runner = RunnerOpts(**{f.name: opt[f.name] for f in fields(RunnerOpts)})
```
Comment (fsoubelet) [inline pylhc_submitter/job_submitter.py:394]: me gusta :)

File: pylhc_submitter/submitter/htc_utils.py [hunk 3]
```diff
@@ -15,35 +15,35 @@
...
 
-EXECUTEABLEPATH = {
-    "madx": MADX_BIN,
-    "python3": PYTHON3_BIN,
-    "python2": PYTHON2_BIN,
-}
-
-
 CMD_SUBMIT = "condor_submit"
 JOBFLAVOURS = (
```
Comment (fsoubelet) [inline pylhc_submitter/submitter/htc_utils.py:48]: Should we take the opportunity to extract these constants to something like `pylhc_submitter.constants.htcondor`?

File: pylhc_submitter/submitter/htc_utils.py [hunk 4]
```diff
@@ -58,25 +58,37 @@
...
-def create_subfile_from_job(cwd: Path, job: str):
-    """Write file to submit to ``HTCondor``."""
+def create_subfile_from_job(cwd: Path, submission: Union[str, htcondor.Submit]) -> Path:
+    """
+    Write file to submit to ``HTCondor``.
+    
+    Args:
+        cwd (Path): working directory
+        submission (str, htcondor.Submit): HTCondor submission definition (i.e. content of the file)
+
```
Comment (fsoubelet) [inline pylhc_submitter/submitter/htc_utils.py:72]: Thanks

File: pylhc_submitter/submitter/htc_utils.py [hunk 5]
```diff
@@ -58,25 +58,37 @@
...
+    Returns:
+        Path: path to sub-file
+
```
Comment (fsoubelet) [inline pylhc_submitter/submitter/htc_utils.py:75]: Empty line here, don't know if it's on purpose but you don't have it in some other functions.

File: pylhc_submitter/submitter/htc_utils.py [hunk 6]
```diff
@@ -58,25 +58,37 @@
...
 
-def submit_jobfile(jobfile: Path, ssh: str):
-    """Submit subfile to ``HTCondor`` via subprocess."""
+def submit_jobfile(jobfile: Path, ssh: str) -> None:
+    """Submit subfile to ``HTCondor`` via subprocess.
+    
+    Args:
+        jobfile (Path): path to sub-file
+        ssh (str): ssh target
+
```
Comment (fsoubelet) [inline pylhc_submitter/submitter/htc_utils.py:90]: Same

File: pylhc_submitter/submitter/htc_utils.py [hunk 7]
```diff
@@ -87,7 +99,16 @@ def submit_jobfile(jobfile: Path, ssh: str):
...
-def _start_subprocess(command):
+def _start_subprocess(command: List[str]) -> int:
+    """ Start subprocess and log output. 
+    
+    Args:
+        command (List[str]): command to execute
+
+    Returns:
+        int: return code of the process
+
```
Comment (fsoubelet) [inline pylhc_submitter/submitter/htc_utils.py:110]: Same

File: pylhc_submitter/submitter/iotools.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+                                                     COLUMN_JOBID, JOBDIRECTORY_PREFIX,
+                                                     JOBSUMMARY_FILE, SCRIPT_EXTENSIONS)
+from pylhc_submitter.submitter import htc_utils
+from pylhc_submitter.submitter.mask import (create_job_scripts_from_mask,
+                                                      generate_jobdf_index, is_mask_file)
+
+LOG = logging.getLogger(__name__)
+
+
+@dataclass
```
Comment (fsoubelet) [inline pylhc_submitter/submitter/iotools.py:27]: 👍

File: pylhc_submitter/submitter/runners.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+def _execute_shell(df_row: pd.Series) -> int:
+    """ Execute the shell script. 
+    
+    Args:
+        df_row (pd.Series): row in the job-dataframe
+    
+    Returns:
+        int: return code of the process
+    """
+    _, column = df_row
```
Comment (fsoubelet) [inline pylhc_submitter/submitter/runners.py:109]: It is a bit unintuitive to name the arg `df_row` when, from looking at i.e. its call in `run_local` above, we are passing the `.iterrows()` which gives a tuple. Maybe we rename this arg?

File: tests/unit/test_job_submitter.py [hunk 10]
```diff
@@ -19,162 +23,258 @@
...
+    """ Tests that wrong URI's are identified. """
+    setup = InputParameters(
+        working_directory=tmp_path, 
+        run_local=True,
+        output_destination="root:/eosuser.cern.ch/eos/my_new_output",
+    )
+    setup.create_mask()
+    with pytest.raises(ValueError) as e:
+        job_submit(**asdict(setup))
+    assert "EOS-URI" in str(e)
```
Comment (fsoubelet) [inline tests/unit/test_job_submitter.py:57]: This tests for a wrong `EOS` uri, maybe we can rename it. How about `test_raises_on_wrong_eos_uri`?

File: tests/unit/test_job_submitter.py [hunk 11]
```diff
@@ -19,162 +23,258 @@
...
+    """ Test that an error is raised if htcondor bindings are not found.
+    If this tests fails, this might mean, that htcondor bindings are finally 
+    available for the other platforms. """
+    setup = InputParameters(working_directory=tmp_path)
+    setup.create_mask()
     with pytest.raises(EnvironmentError) as e:
-        job_submit(**setup)
-    assert "htcondor bindings" in e.value.args[0]
+        job_submit(**asdict(setup))
+    assert "htcondor bindings" in str(e)
```
Comment (fsoubelet) [inline tests/unit/test_job_submitter.py:115]: I would take the opportunity to rename this guy too, maybe `test_nonlinux_run_raises` or something

**Review by fsoubelet — COMMENTED** [review]
Some comments here and there. I like the clearer structure and addition of many docstrings.

I suggest updating the `CHANGELOG` file too before we merge.

File: doc/conf.py [hunk 12]
```diff
@@ -31,20 +31,10 @@
...
-        entry.split(" = ")[0]: entry.split(" = ")[1].strip('"')
-        for entry in about_text.strip().split("\n")
-        if entry.startswith("__")
-    }
 
-
-ABOUT_PYLHC_SUBMITTER = about_package(ABOUT_FILE)
+ABOUT_PYLHC_SUBMITTER: dict = {}
+with ABOUT_FILE.open("r") as f:
+    exec(f.read(), ABOUT_PYLHC_SUBMITTER)
```
[the remark this answers is not in the document]
Comment (JoschD) [reply doc/conf.py:37]: Sorry. I copied this from `omc3` but I guess it was outdated there.

File: pylhc_submitter/job_submitter.py (hunk 1 above)
Comment (JoschD) [reply pylhc_submitter/job_submitter.py:330]: This is on purpose. Two reasons:
a) I want the ini-file as close to the user-input as possible (even with mistakes).
b) the `check_opts` does not give the checked opts back, as it transforms it into the creation/runner options. So they are not even available for saving anymore

File: pylhc_submitter/submitter/runners.py (hunk 9 above)
Comment (JoschD) [reply pylhc_submitter/submitter/runners.py:109]: indeed. my mistake

File: pylhc_submitter/job_submitter.py (hunk 1 above)
Comment (fsoubelet) [reply pylhc_submitter/job_submitter.py:332]: Okay. That second point escaped me, I'm fine with all this.

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,5 +1,18 @@
 # `pylhc-submitter` Changelog
 
+## Version 2.0.0
+
+- General code cleanup/refactoring/documentation.
+ - partly breaks backward compatibility, if individual methods of the `job_submitter`-functionality have been used.
+ - does not affect any setups simply calling the `main()` function of `job_submitter.py` or calling the `job_submitter` as a module.
```
Comment (fsoubelet) [inline CHANGELOG.md:7]: Unaligned `-` here and not capitalized sentences :P

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,5 +1,18 @@
...
+## Version 2.0.0
+
+- General code cleanup/refactoring/documentation.
+ - partly breaks backward compatibility, if individual methods of the `job_submitter`-functionality have been used.
+ - does not affect any setups simply calling the `main()` function of `job_submitter.py` or calling the `job_submitter` as a module.
+ - Apart from some fixed imports, following the new structure, the `autosix` module has been untouched. 
+
+
+- New Feature of `job_submitter`:
+  - `output_destination` input parameter, which sets an output directory in which the folder-stucture
```
Comment (fsoubelet) [inline CHANGELOG.md:12]: same

**Comment by JoschD** [comment]
@anabramo last chance to give it a look, before it will be merged ;)

**Review by fsoubelet — APPROVED** [review]

**Comment by anabramo** [comment]
Look great to me!

**Pull request merged**